cluster-autoscaler/cloudprovider/azure/azure_vms_pool.go

```diff
@@ -102,28 +120,260 @@ func (agentPool *VMsPool) MaxSize() int {
+	return size, nil
 }
 
-// IncreaseSize increase the size through a PUT AP call. It calculates the expected size
-// based on a delta provided as parameter
+// IncreaseSize increase the size through a PUT AP call.
 func (agentPool *VMsPool) IncreaseSize(delta int) error {
-	// TODO(wenxuan): Implement this method
-	return cloudprovider.ErrNotImplemented
+	if delta <= 0 {
+		return fmt.Errorf("size increase must be positive, current delta: %d", delta)
+	}
+
+	currentSize, err := agentPool.getVMsPoolSize()
+	if err != nil {
+		return err
+	}
+
+	if currentSize == -1 {
+		return fmt.Errorf("the vms pool %s is still under initialization, skip the size increase", agentPool.Name)
+	}
+
+	if int(currentSize)+delta > agentPool.MaxSize() {
+		return fmt.Errorf("size-increasing request of %d is bigger than max size %d", int(currentSize)+delta, agentPool.MaxSize())
+	}
+
+	return agentPool.scaleUpToCount(currentSize + int64(delta))
+}
+
+func (agentPool *VMsPool) buildRequestBodyForScaleUp(count int64) (armcontainerservice.AgentPool, error) {
+	versionedAP, err := agentPool.getAgentpool()
+	if err != nil {
+		klog.Errorf("Failed to get vms pool %s, error: %s", agentPool.Name, err)
+		return armcontainerservice.AgentPool{}, err
+	}
+
+	if versionedAP.Properties != nil &&
+		versionedAP.Properties.VirtualMachinesProfile != nil &&
+		versionedAP.Properties.VirtualMachinesProfile.Scale != nil {
+		requestBody := armcontainerservice.AgentPool{
+			Properties: &armcontainerservice.ManagedClusterAgentPoolProfileProperties{
+				Type: versionedAP.Properties.Type,
+			},
+		}
+
+		// the request body must have the same mode as the original agentpool
+		// otherwise the PUT request will fail
+		if versionedAP.Properties.Mode != nil &&
+			*versionedAP.Properties.Mode == armcontainerservice.AgentPoolModeSystem {
+			systemMode := armcontainerservice.AgentPoolModeSystem
+			requestBody.Properties.Mode = &systemMode
+		}
+
+		// self-hosted CAS will be using Manual scale profile
+		if len(versionedAP.Properties.VirtualMachinesProfile.Scale.Manual) > 0 {
+			// set the count for first manual scale profile to the new target value
+			virtualMachineProfile := *versionedAP.Properties.VirtualMachinesProfile
+			versionedAP.Properties.VirtualMachinesProfile.Scale.Manual[0].Count = to.Int32Ptr(int32(count))
+			requestBody.Properties.VirtualMachinesProfile = &virtualMachineProfile
+			return requestBody, nil
+		}
+		// aks-managed CAS will be using AutoscaleProfile
+		// TODO(wenxuan): support AutoscaleProfile once it become available with the new API release
+	}
+	return armcontainerservice.AgentPool{}, fmt.Errorf("failed to build request body for scale up, agentpool doesn't have valid virtualMachinesProfile")
+}
+
+// scaleUpToCount sets node count for vms agent pool to target value through PUT AP call.
+func (agentPool *VMsPool) scaleUpToCount(count int64) error {
```

Comment:
I generally prefer always-one-use methods like this to be a part of the code calling it, rather than being a separate method, especially when there are no clear boundary of responsibility. I think the case of `IncreaseSize` and `scaleUpToCount` (and more) applies.
In case you have a similar opinion, I don't think you need to be restricted by the pattern in `azure_scale_set.go` for this case.

Reply:
I don't have a preference here, just thought it might make it a bit easier for maintainers if it has the same pattern with vmss. I would like to collect more feedback of this PR and make changes together if needed.